### PR TITLE
Improve Cashbox Report

### LINKED ISSUE #3595: Rewrite Cash Report
The Cash Report provides a simplistic view of the cashbox as a series of entries and exits.  This is helpful for non-accountants when reviewing their daily reports to compare to the system.

However, the report is very outdated.  [Here is the server-side controller](https://github.com/IMA-WorldHealth/bhima/blob/master/server/controllers/finance/reports/cashReport/index.js).  There are a number of issues with it:
 1. It should use period totals to establish an opening balance and running balance.
 2. The opening balance is not in the cashbox's or enterprise's currency - it is in USD.
 3. The report should _optionally_ include unposted records, and should provide an italics representation like the account report.

There is also a bug where the dates are not respected.

### PR DESCRIPTION
**Pull request by jniles**
This PR improves the cashbox report in the following ways:
 1. The dates (from - to) are now respected in the search query.
 2. The currencies are rendered in the correct cashbox currency.
 3. We include a real opening balance and closing balance.
 4. The user can identify the difference between posted and unposted records via italics applied to unposted records.

We've also centralized the `getAccountTransactions()` call used between multiple reports into a single shared file `accounts/transactions`.  We've also rewritten the query for the combined posting_journal and general_ledger such that the WHERE statement is executed in subqueries, greatly improving the performance of the combined query.

Finally, the `getAccountTransactions()` function now sorts first based on `created_at` date, then on `trans_date`.  This makes back-dated transactions show up later in the rendering.  For example, users frequently back-date the transfer from the cashbox into the main coffre, and it makes sense to show these at the bottom of the report.

Closes #3595.

### CODE REVIEW

**Comment by jniles** [comment]
@mbayopanda, can I get a review?

File: server/controllers/finance/accounts/transactions.js [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+ *  2. Exchange Rate Calculation
+ *  3. Running Balance Calculation
+ *
+ */
+
+const _ = require('lodash');
+const db = require('../../../lib/db');
+const FilterParser = require('../../../lib/filter');
+const Exchange = require('../exchange');
+const Accounts = require('.');
```
Comment (mbayopanda) [inline server/controllers/finance/accounts/transactions.js:17]: What does it means ? The current file or all files of the current directory.

If it is the current file, why do we need to import it ?

File: server/controllers/finance/accounts/transactions.js [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+  const filters = new FilterParser(options);
+
+  options.isEnterpriseCurrency = options.isEnterpriseCurrency || false;
+
+  // return the proper way to convert the values depending on if we are in the
+  // exchange rate currency or not.
+  const enterpriseCurrencyColumns = `
+    (debit / rate) AS exchangedDebit, (credit / rate) AS exchangedCredit,
+    (balance / rate) AS exchangedBalance, @cumsum := (balance / rate) + @cumsum AS cumsum
+  `;
```
Comment (mbayopanda) [inline server/controllers/finance/accounts/transactions.js:43]: The request is so clear 👍

**Review by mbayopanda — COMMENTED** [review]
LGTM 👍 

I was not able to test it with real data of imck because the cashbox selection did not work

File: server/controllers/finance/accounts/transactions.js (hunk 1 above)
Comment (jniles) [reply server/controllers/finance/accounts/transactions.js:17]: Aha.  That is confusing.  It actually imports `./index.js`.  This file is `./transactions.js`.

**Comment by jniles** [comment]
@mbayopanda, thank you for thoroughly testing this module.  It appears that the `bhCashboxSelect` component [defaults to `restrict-to-user="true"`](https://github.com/IMA-WorldHealth/bhima/blob/master/client/src/js/components/bhCashboxSelect.js#L36) and so we couldn't see the data due to not having the correct permissions.

I've fixed this by turning the restriction off.

**Pull request merged**